R/raincloudPlots.R

```diff
@@ -0,0 +1,908 @@
+      order = 1, reverse = options$horizontal, override.aes = list(alpha = 0.5, color = NA)  # NA removes points in boxes
+    )
+  }
+  guideColor <- if (options$covariate == "") {
+    NULL
+  } else {
+    if (is.factor(dataset$covariate)) {
+      ggplot2::guide_legend(override.aes = list(size = 6.5))
+    } else {
+      ggplot2::guide_colourbar()
+    }
+  }
+  guide <- ggplot2::guides(fill = guideFill, color = guideColor)
+
+  legendCloser <- ggplot2::theme(legend.box.spacing = ggplot2::unit(0, "pt"), legend.margin = ggplot2::margin(0, 0, 0, 0))
+
+  plotInProgress <- plotInProgress + guide + legendCloser
+
+  # Caption
+  if (options$showCaption) {
+    caption         <- .rainCaption(options, dataInfo, dataInfo$intervalBounds[[inputVariable]], warningAxisLimits, errorVioSides)
+    addCaption      <- ggplot2::labs(caption = caption)
+    captionPosition <- ggplot2::theme(plot.caption = ggtext::element_markdown(hjust = 0))  # Bottom left position
+    plotInProgress  <- plotInProgress + addCaption + captionPosition
+  }
+
+  # Assign to inputPlot
+  return(plotInProgress)
+}  # End .rainFillPlot()
+
+
+
+# .rainSetPalettes() ----
+.rainSetPalettes <- function(dataset, options) {
+
+  fillTitle <- if (options$secondaryFactor == "") options$primaryFactor else options$secondaryFactor
+  paletteFill <- if (options$secondaryFactor != "" || options$colorAnyway) {
+    jaspGraphs::scale_JASPfill_discrete(options$colorPalette, name = fillTitle)
+  } else {
+    NULL
+  }
+
+  paletteColor <- if (options$covariate != "") {
+    if (is.factor(dataset$covariate)) {
+      jaspGraphs::scale_JASPcolor_discrete(  options$covariatePalette, name = options$covariate)
+    } else {
+      jaspGraphs::scale_JASPcolor_continuous(options$covariatePalette, name = options$covariate)
+    }
+  } else {
+    if (options$secondaryFactor != "" || options$colorAnyway) {
+      jaspGraphs::scale_JASPcolor_discrete(  options$colorPalette,     name = options$secondaryFactor)
+    } else {
+      NULL
+    }
+  }
+
+  return(list(fill = paletteFill, color = paletteColor))
+}  # End .rainSetPalettes()
+
+
+
+# .rainInfoFactorCombinations() ----
+# Calculates info to determine colors & geom orientation
+.rainInfoFactorCombinations <- function(inputDataset, inputPlot, extractColor = TRUE) {
+  onlyFactors    <- inputDataset[c("primaryFactor", "secondaryFactor")]
+  # Extract used Fill colors from plot
+  # https://stackoverflow.com/questions/11774262/how-to-extract-the-fill-colours-from-a-ggplot-object
+
+  if (extractColor) {  # So that we can re-use it in .rainFillTable()
+
+    onlyFactors$color <- tryCatch(
+      {
+        ggplot2::ggplot_build(inputPlot)$data[[1]]["fill"]$fill  # Requires secondaryFactor or colorAnyway
+      },
+      error = function(e) {                                      # Thus error handling, but not used further
+        return("black")
+      },
+      warning = function(w) {
+        return("black")
+      }
+    )
```

Comment:
This looks a bit sus, why do we need this? Aren't there other ways to determine this without building the plot?

Reply:
This was my rationale:

Goal:
points, vioOutline, boxOutline, mean, meanLines should have the option to be shown in the group color according to the colorPalette.

Required info:
group color(s) according to the colorPalette

Problem:
Depending on the number of groups,
`jaspGraphs::scale_JASPfill_discrete()` and `jaspGraphs::scale_JASPcolor_discrete()`
as applied to a ggplot
will yield different hexcodes. If you have two (vs. three) groups, then you get two (vs. three) hexcodes.
**However**, the first two hexcodes did not always match when I tried this out (maybe this is also specific to some of the colorPalettes?).

Solution:
First apply the colorPalette (section Palettes in `.rainFillPlot()`), then extract the assigned hexcodes through `ggplot_build()` (see `.rainInfoFactorCombinations()`, Preparation section in `.rainFillPlot()`).

Because not every plot will have a fill (requires secondaryFactor or colorAnyway),
it will not always be possible to extract the hexcodes. This is why I used `tryCatch()`.
I guess I could do an if-else statement instead of `tryCatch()` (might be a bit cleaner?).
But as I remember it, I found no work-around and ended up having to `ggplot_build()`.
Because only then you have access to the group colors from the colorPalette in the way ggplot2 decided to assign them to the groups.

Please let me know, what you think! 😄

Reply:
it indeed looks sus and I'm not sure I really comprehend the explanation why we need tryCatch -- I would reserve that function for actual error catching and use if else for all other purposes

Reply:
I already fixed this with the help of Don, but did not document it here further.
Now, we simply extract as many hexcodes as there are levels in the factor, without building the plot:

<img width="991" alt="Screenshot 2024-04-15 at 16 16 53" src="https://github.com/jasp-stats/jaspDescriptives/assets/123652963/35bb7c50-36f3-4c97-b816-8535a9f560f3">